File: modules/x-engine-native-direct_lrn/iOS/Class/Native_direct_lrn.m
//
//  Native_direct_lrn.m
//  direct_lrn
//
//  Created by zk on 2020/9/7.
//  Copyright © 2020 edz. All rights reserved.


#import "Native_direct_lrn.h"
#import "XENativeContext.h"
#import "Unity.h"
#import "NSURL+QueryDictionary.h"
#import <x-engine-native-core/Unity.h>
#import "ReactNativeViewController.h"

@interface Native_direct_lrn()
@property (nonatomic, strong) id<iDirect>  lrnDirect;
@end

@implementation Native_direct_lrn
NATIVE_MODULE(Native_direct_lrn)

- (NSString*) moduleId{
    return @"com.zkty.native.direct_lrn";
}
- (nonnull NSString *)protocol {
    return @"lrn:";
}
-(NSString*) scheme{
    return @"lrn";
}

- (int) order{
    return 0;
}

- (void)afterAllNativeModuleInited{
}

- (nonnull UIViewController *)getContainer:(nonnull NSString *)protocol host:(nullable NSString *)host pathname:(nonnull NSString *)pathname fragment:(nullable NSString *)fragment query:(nullable NSDictionary<NSString *,id> *)query params:(nullable NSDictionary<NSString *,id> *)params frame:(CGRect)frame   {
    
    if(!protocol){
        protocol = [self protocol];
    }
    
    id lrn_params = params[@"__RN__"];
    NSAssert(lrn_params  , @"__RN__ 不存在");
    if(!lrn_params) return nil;
    
    id moduleName = lrn_params[@"moduleName"];
    NSAssert(moduleName  , @"moduleName 不存在");
    if(!moduleName) return nil;

    
    BOOL isHideNavBar = [params[@"hideNavbar"] boolValue];
    NSString *finalUrl = @"";
    
    
    NSAssert(!fragment || ![fragment hasPrefix:@"#"]  , @"fragment 不需要加#");
    fragment = (fragment && fragment.length>0) ? [NSString stringWithFormat:@"#%@",fragment] : @"";
    NSString* queryStr   = query.uq_URLQueryString && query.uq_URLQueryString.length>0 ? [NSString stringWithFormat:@"?%@",query.uq_URLQueryString]:@"";
    
    finalUrl = [NSString stringWithFormat:@"%@//%@%@%@%@",protocol,host,pathname,fragment,queryStr];
    
    ReactNativeViewController *vc =  [[ReactNativeViewController alloc] initWithUrl:finalUrl withHiddenNavBar:isHideNavBar webviewFrame:frame moduleName:moduleName];
    vc.hidesBottomBarWhenPushed = YES;
    return vc;
}

@end

File: modules/x-engine-native-direct_lrn/android/native-direct_lrn/src/main/java/com/zkty/nativ/direct_lrn/Nativedirect_lrn.java
package com.zkty.nativ.direct_lrn;

import android.app.Activity;
import android.util.Log;

import com.tencent.mmkv.MMKV;
import com.zkty.nativ.core.NativeModule;
import com.zkty.nativ.core.XEngineApplication;
import com.zkty.nativ.direct.IDirect;
import com.zkty.nativ.react_native.XEngineReactNativeActivity;

import java.io.File;
import java.util.Map;

public class Nativedirect_lrn extends NativeModule implements IDirect {

    @Override
    public String moduleId() {
        return  "com.zkty.native.direct_lrn";
    }

    @Override
    public void afterAllNativeModuleInited() {

    }

    @Override
    public String scheme() {
        return "lrn";
    }

    @Override
    public String protocol() {
        return "file://";
    }

    @Override
    public void push(String protocol, String host, String pathname, String fragment, Map<String, String> query, Map<String, String> params) {
        Activity mActivity = XEngineApplication.getCurrentActivity();
        if(mActivity == null){
            return;
        }
        //应用名称
        String moduleName = params.get("moduleName");
        //本地 jsBundle 名称
        String bundleAssetName = params.get("bundleAssetName");
        //debug 本地 jsBundle Serve  jsMainModulePath 设置
        String jsMainModulePath = params.get("jsMainModulePath");

        XEngineReactNativeActivity.start(mActivity,moduleName,bundleAssetName,jsMainModulePath,query);
    }

    @Override
    public void back(String host, String fragment) {

    }
}
